test/jdk/javax/crypto/KeyGenerator/CompareKeys.java

```diff
@@ -0,0 +1,147 @@
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+import javax.crypto.SecretKey;
+import javax.crypto.spec.SecretKeySpec;
+import java.security.Key;
+import java.security.KeyFactory;
+import java.security.KeyPair;
+import java.security.KeyPairGenerator;
+import java.security.Provider;
+import java.security.Security;
+import java.security.spec.PKCS8EncodedKeySpec;
+import java.security.spec.X509EncodedKeySpec;
+import java.util.Arrays;
+import java.util.LinkedList;
+import java.util.List;
+import javax.crypto.KeyGenerator;
+
+/*
+ * @test
+ * @bug 8185127
+ * @summary Test key comparison for the Keys generated through KeyGenerator
+ * @run main CompareKeys
+ */
+public class CompareKeys {
+
+    public static void main(String[] args) throws Exception {
+
+        for (KeygenAlgo alg : getSupportedAlgo("KeyGenerator")) {
+            System.out.printf("Verifying provider %s and algorithm %s%n",
+                    alg.provider().getName(), alg.algoName());
+            SecretKey k = genSecretKey(alg.algoName(), alg.provider());
+            checkKeyEquality(k, copy(alg.algoName(), k));
+        }
+
+        for (KeygenAlgo alg : getSupportedAlgo("KeyPairGenerator")) {
+            System.out.printf("Verifying provider %s and algorithm %s%n",
+                    alg.provider().getName(), alg.algoName());
+            KeyPair kp = genKeyPair(alg.algoName(), alg.provider());
+            checkKeyPairEquality(kp, copy(alg.algoName(), kp));
+        }
+        System.out.println("Done!");
+    }
+
+    @SuppressWarnings("preview")
+    private record KeygenAlgo(String algoName, Provider provider) {
+
+    }
+
+    private static void checkKeyPairEquality(KeyPair origkp, KeyPair copykp)
+            throws Exception {
+
+        checkKeyEquality(origkp.getPrivate(), copykp.getPrivate());
+        checkKeyEquality(origkp.getPublic(), copykp.getPublic());
+    }
+
+    /**
+     * Compare original Key with another copy.
+     */
+    private static void checkKeyEquality(Key origKey, Key copyKey) {
+
+        if (!(origKey.equals(copyKey)
+                || origKey.hashCode() == copyKey.hashCode())
+                && !Arrays.equals(origKey.getEncoded(), copyKey.getEncoded())
+                && !origKey.getFormat().equals(copyKey.getFormat())) {
```

Comment:
So, all of these 3 checks have to fail in order to be considered key inequality? Could you spell out clearly what is expected here? I am not sure if this compound condition is correct. As it is now, the copy must have different format(2nd condition) AND different encoding(3rd condition) AND (not equals AND not same hash code)(1st condition), in order to trigger the RuntimeException.